Use Swift.Error instead of NSError

ffried opened the pull request:
This introduces a new kind of error handling by completely using Swift's `Error` for all errors. This allows a more typesafe access to error information. Using the newly introduced `ErrorInformation` struct together with `ErrorInformationKey` it’s still possible to store an arbitrary amount of additonal error data, however, and still be completely typesafe.
Operations are now encouraged to implement their own `Error` Type instead of relying on `NSError`. As an example, `TimeoutObserver` now defines its own error.

ffried commented:
This PR is in favor of #86 which I'll close now.

jjatie commented on PSOperations/Capability.swift:
```diff
@@ -8,15 +8,53 @@
+                    return true
+            case (.failed(_ /*let lhsError*/), .failed(_ /*let rhsError*/)):
+                return true
+                // We could compare based on the errors bridged to NSError. But it would be rather unsafe.
```
Why not make a separate case just for this?
```
        case (.failure(let l as NSError), .failure(let r as NSError)):
            return l == r

        case (.failure, .failure):
            return true
```
Alternatively you could make the Error generic
```
enum Reason<E: Error> {
    case notDetermined
    case notAvailable
    case denied
    case failed(E)
}

extension Reason: Equatable where E: Equatable {
    static func ==(lhs: Reason, rhs: Reason) -> Bool {
        switch (lhs, rhs) {
        case (.notDetermined, .notDetermined),
             (.notAvailable, .notAvailable),
             (.denied, .denied):
            return true

        case (.failed(let l), .failed(let r)):
            return l == r

        default:
            return false
        }
    }
}
```

ffried replied:
Because all Swift `Error`s are bridgeable to `NSError`.

jjatie replied:
If they are all bridgeable, they can all be equated, can they not?

ffried replied:
Because `NSError` does not e.g. use a custom `Equatable` implementation. It only checks the type.
```swift
struct MyCustomError: Error, Equatable {
    let someInt: Int
    let someString: String
}

let error1 = MyCustomError(someInt: 4, someString: "ABC")
let error2 = MyCustomError(someInt: 3, someString: "DEF")

error1 == error2 // false
(error1 as NSError) == (error2 as NSError) // true
```

ffried replied:
Resolved by removing the `Equatable` conformance.

jjatie commented:
I'm not sure what the benefit of `ErrorInformation` is. The user is still going to need to know all of the custom keys associated with a given error. It's a neat way of making a type-safe user info dictionary, but why not just implement custom errors that have properties for each key? A generic `ConditionError` conforming to `CapabilityError` makes more sense to me:
```
public protocol ConditionError: Error {
    var conditionName: String { get }
}

public enum CapabilityError<C: Capability>: CapabilityError {
    var conditionName: String {
        return "\(C.self)"
    }

    case notDetermined
    case notAvailable
    case denied
    case failed(Error)

    init?(status: CapabilityStatus) {
        switch status {
        case .authorized: return nil
        case .notDetermined: self = .notDetermined
        case .notAvailable: self = .notAvailable
        case .denied: self = .denied
        case .error(let e): self = .error(e)
    }
}
```

`CapabilityError` is much smaller here and encapsulates all the same information. Additionally, you don't need to fall back to string pattern matching and instead have strict types for checking error types i.e instead of this:
```
do {
     try errorThrowingCapabilityError()
} catch let e as CapabilityError {
    switch e.capabilityName {
    case "capabilityName1": // Even if these are type-safe keys, it's not very nice.
    case "capabilityName2":
    ....
}
```
You can write:
```
do {
     try errorThrowingCapabilityError()
} catch let e as CapabilityError<Capability1> {
    //...
} catch let e as CapabilityError<Capability2> {
    //...
}
```
Adding a `typealias` for your capabilities can make it a little nicer too 
```
typealias Capability1Error = CapabilityError<Capability1>
do {
     try errorThrowingCapabilityError()
} catch let e as Capability1Error {
    //...
}
```
You can also just catch all `ConditionError`s in the same manner.
```
do {
     try errorThrowingCapabilityError()
} catch let e as ConditionError {
    //...
}
```

Errors for other operations, like `NoCancelledDependencies`, are simply implemented as:
```
struct CancelledDependenciesError: ConditionError {
    let conditionName = "CancelledDependencies"
    let cancelledDependency: Operation
}
```
It's 100% type-safe; you pattern match on the type, not a string or key, and the properties associated with it are type-safe and easily discoverable through autocomplete.

ffried commented:
Not everything you mentioned is actually possible, since you can't use `ConditionError` as existential if it inherits from `Equatable` due to the `Self` requirements.

But your idea to have concrete errors for each condition is nevertheless great and I've implemented this now.
Also, I've removed the `Equatable` conformances. It wasn't used inside PSOperations. Also, it's easy enough to check for a given case with e.g. `if case .notAvailable = status` or a `switch`. I think this is overall the better solution.

jjatie commented on PSOperations/Capability.swift:
```diff
@@ -82,7 +91,7 @@ public struct Capability<C: CapabilityType>: OperationCondition {
 }
 
-private class AuthorizeCapability<C: CapabilityType>: Operation {
+fileprivate class AuthorizeCapability<C: CapabilityType>: Operation {
```
Why bother changing this? `private` at the file scope is `fileprivate`

ffried commented on PSOperations/OperationCondition.swift:
```diff
@@ -45,15 +42,27 @@ public protocol OperationCondition {
     func evaluateForOperation(_ operation: Operation, completion: @escaping (OperationConditionResult) -> Void)
 }
 
+public protocol AnyConditionError: Error {
```
The name is IMHO necessary, because for the protocol `AnyConditionError`, there's no way to find out what condition it was. And we need a protocol that can be used as existential for e.g. `OperationConditionResult`. We also can't make the latter generic, since the type of the Error is not always known. Furthermore, then `OperationCondition` itself would need to become generic which would complicate a lot more things.

jjatie replied:
D'oh! Forgot `OperationCondition` is a protocol

jjatie commented:
This looks great! Any idea on how to grab attention from PluralSight? Seems like they haven't touched it in about 6 months other than to push an update for the new tools.

ffried commented:
@jjatie Sadly no. In some projects we opted for using our fork (from which this PR originates). This project could seriously need some more collaborators with write access that actually care.

jjatie commented:
@ffried FYI the tests on your branch failed:

> Testing failed:
> 	“Swift Language Version” (SWIFT_VERSION) is required to be configured correctly for targets which use Swift. Use the [Edit > Convert > To Current Swift Syntax…] menu to choose a Swift version or use the Build Settings editor to configure the build setting directly.
> ** TEST FAILED **
>

ffried commented:
@jjatie That's odd. They've just passed here?
<img width="1167" alt="image" src="https://user-images.githubusercontent.com/1689782/48477674-7b3d6580-e802-11e8-89ce-ba684e9fe9a4.png">

jjatie commented:
@ffried Weird! I was just looking at the status in your fork.

ffried commented:
@jjatie Which status?

jjatie commented:
@ffried The one reported by GitHub when clicking on [branches](https://github.com/sersoft-gmbh/PSOperations/branches)

ffried commented:
Oh, I see. Didn't know travis was running against this. The config is way out of date (Xcode 8.3). I've just updated it. Should probably pass now.

ffried commented:
Alright, the Apple TV build wouldn't connect to the simulator, so I disabled it. The build passes now.

mcmurrym commented:
This branch is about Errors, not a out disabling tests for a platform. That issue needs to be fixed elsewhere

mcmurrym commented:
like wise the travis fixes should happen else where. I'll read up on this pull request today.

ffried commented:
@mcmurrym Good point. I'll just remove the last two commits and put them into a separate PR.

ffried commented:
See #116.

jjatie commented:
@mcmurrym Any update on when this can be merged? Maybe with Swift 5?
@ffried There are now merge conflicts

ffried commented:
@jjatie I've fixed the conflicts.
